keps/sig-apps/4026-crojob-scheduled-timestamp-annotation/README.md

```diff
@@ -0,0 +1,290 @@
+- [X] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Currently, there is no supported way to get the original/expected initial scheduled timestamp for the job created from a cronjob. This KEP proposes to set the original scheduled time as an annotation in the job metadata.
+
+## Motivation
+
+### Goals
+
+- Set job scheduled timestamp as an annotation on the job.
+- Adding the annotation should not be disruptive to existing workloads.
+
+### Non-Goals
+
+## Proposal
+
+At a high level, the proposal is to modify the CronJob controller to set the job scheduled timestamp as a job annotation. The details of this are outlined in the Design Details section below.
+
+Job scheduled timestamp annotation: `batch.kubernetes.io/cronjob-scheduled-timestamp`
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a user, I would like to get the job's scheduled timestamp that this job was expected to be running.
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+CronJobs are always working with the assumption that the changes apply only to newly created jobs after the change. Therefore, the change will be to inject the annotation for newly created Jobs from CronJobs for when the feature is on. This will nicely play with downgrade and doesn't introduce unnecessary complexity.
+
+## Design Details
+
+The CronJob controller will only need a minor update to the [getJobFromTemplate2](https://github.com/kubernetes/kubernetes/blob/7024beeeeb1f2e4cde93805a137cd7ad92fec466/pkg/controller/cronjob/utils.go#L188) function, to add the job scheduled timestamp as the job annotation `batch.kubernetes.io/cronjob-scheduled-timestamp`. The scheduled timestamp is represented in `RFC3339`.
+
+For the scheduled timestamp's timezone, the initial thought was to use `UTC` as it's used as the primary one for less confusion. However, since the `job` object has a `spec.timeZone`, it was a better to use the same timezone within the same object. If the job `spec.timeZone` is not set or `nil`, the annotation will use the `UTC` timezone as a default.
```

Comment:
How do we know that UTC is right? The actual timezone that the CronJob controller uses is the system (might be a container, not the entire host) timezone as seen by kube-controller-manager.

Reply:
So long as we use the time _in UTC_ (ie, not using any configured local time that drives the actual scheduling of new Jobs), then I think this is fine.